[Benchmarks] Change all args to from `no_...` to their positive form

fmcurti opened the pull request:
Fixes #7072

As @stas00 pointed out a while back, there were a bunch of negative args formulations in `benchmarks.py` which are hard to understand and make the code unnecessarly complex.

This PR fixes this problem by 
a) Formulating all args in their positive form 
b) Improving their docstring
c) Making the code more readable

**!!! This has breaking changes for the command line args of `examples/run_benchmark.py `and `examples/run_benchmark_tf.py` !!!**
People that were using `--no_...` on command line previously, now should use `--no-...`. *E.g.* `--no_inference` is changed to `--no-inference` on the command line.

Thanks a lot @fmcurti for leading this PR!

patrickvonplaten commented on src/transformers/benchmark/benchmark.py:
```diff
@@ -229,7 +229,7 @@ def _measure_memory(self, func: Callable[[], None]) -> [Memory, MemorySummary]:
                 # tpu
                 raise NotImplementedError(
-                    "Memory Benchmarking is currently not implemented for TPU. Please disable memory benchmarking with `--no_memory` or `args.no_memory=True`"
+                    "Memory Benchmarking is currently not implemented for TPU. Please disable memory benchmarking with `args.memory=False`"
```
```suggestion
                    "Memory Benchmarking is currently not implemented for TPU. Please disable memory benchmarking with `--no-memory` or `args.memory=False`"
```

patrickvonplaten commented on src/transformers/benchmark/benchmark_args_utils.py:
```diff
@@ -57,22 +57,22 @@ class BenchmarkArguments:
-    no_inference: bool = field(default=False, metadata={"help": "Don't benchmark inference of model"})
-    no_cuda: bool = field(default=False, metadata={"help": "Whether to run on available cuda devices"})
-    no_tpu: bool = field(default=False, metadata={"help": "Whether to run on available tpu devices"})
+    inference: bool = field(default=True, metadata={"help": "Whether to benchmark inference of model"})
```
I like this text! Could you add the following sentence to all `default=True` args: "inference can be disabled via `--no-inference.`" So that people know how the param can be set to `False` via the command line

patrickvonplaten commented on src/transformers/benchmark/benchmark_args_utils.py:
```diff
@@ -57,22 +57,22 @@ class BenchmarkArguments:
     log_print: bool = field(default=False, metadata={"help": "Save all print statements in a log file"})
-    no_env_print: bool = field(default=False, metadata={"help": "Don't print environment information"})
-    no_multi_process: bool = field(
+    env_print: bool = field(default=False, metadata={"help": "Wheter to  print environment information"})
```
```suggestion
    env_print: bool = field(default=False, metadata={"help": "Whether to print environment information"})
```

patrickvonplaten commented:
Hey @fmcurti - this looks great, thanks a lot for your contribution here!

In order to pass the `check_code_quality` test it would be great if you can run `make style` as described in https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md . 
And it would be important to make the benchmarks.py test pass :-) 

Let me know if you are stuck and need help :-)

fmcurti commented:
Hi @patrickvonplaten =) 
I think I modified the tests so that they pass, but i think the tests running on circleci are the ones on the master branch and not on my commit maybe? because the error messages I see seem to be from things I already changed

patrickvonplaten commented on src/transformers/benchmark/benchmark_args_utils.py:
```diff
@@ -1,131 +1,147 @@
+    log_print: bool = field(default=False, metadata={"help": "Save all print statements in a log file"})
+    env_print: bool = field(default=False, metadata={"help": "Whether to print environment information"})
+    multi_process: bool = field(
+        default=False,
```
```suggestion
        default=True,

```

patrickvonplaten commented on src/transformers/benchmark/benchmark_args_utils.py:
```diff
@@ -1,131 +1,147 @@
+    multi_process: bool = field(
+        default=False,
+        metadata={
+            "help": "Whether to use multiprocessing for memory and speed measurement. It is highly recommended to use multiprocessing for accurate CPU and GPU memory measurements. This option should only be used for debugging / testing and on TPU."
```
```suggestion
            "help": "Whether to use multiprocessing for memory and speed measurement. It is highly recommended to use multiprocessing for accurate CPU and GPU memory measurements. This option should only be disabled for debugging / testing and on TPU."

```

patrickvonplaten commented on src/transformers/benchmark/benchmark_args_utils.py:
```diff
@@ -1,131 +1,147 @@
+    memory: bool = field(
+        default=True,
+        metadata={
+            "help": "Whether to perform memory measurements. Memory measurements can be disabled via --no-speed"
```
```suggestion
            "help": "Whether to perform memory measurements. Memory measurements can be disabled via --no-memory"

```

patrickvonplaten commented:
@LysandreJik @sgugger - For some reason the git-diff is messed up in this PR. I checked the changes offline using `vimdiff` and the changes look good to me! 
Think the breaking changes are not super relevant here because it's only the benchmarking argument class that is affected and not too many people seem to be using the benchmarking tools.

sgugger reviewed (approved):
Looks good to me, thanks for taking care of this!
Test failure is fixed on master, so this should disappear once merged.

fmcurti commented:
Great! My first contribution to an open source library 🤗
Small one but it's a start,

LysandreJik reviewed (commented):
Is it impossible to keep backwards compatibility on arguments? I really think we should aim for 0 breaking changes to the library's internals, irrespective of usage, and especially for a cosmetic change.

stas00 commented:
The first post looks confusing, it looks like someone edited it, but didn't indicate so, so it appears that @fmcurti praises himself ;) perhaps a ----- bar and a note that it was edited?

stas00 commented:
I'm in agreement with @patrickvonplaten - I don't think this API is being used much and it's an important change for the long run, so it's worth the breaking. Good APIs are not easy.

stas00 commented:
> @LysandreJik @sgugger - For some reason the git-diff is messed up in this PR. I checked the changes offline using `vimdiff` and the changes look good to me!

You need to tell github to ignore whitespaces and then it works:
https://github.com/huggingface/transformers/pull/7075/files?diff=split&w=1
Got it via this:
![screenshot_6](https://user-images.githubusercontent.com/10676103/93915406-d2dd6f80-fcbc-11ea-87fa-2bfbbabbed5b.png)

LysandreJik commented:
> I'm in agreement with @patrickvonplaten - I don't think this API is being used much and it's an important change for the long run, so it's worth the breaking. Good APIs are not easy.
>
> but if it's really needed one could keep the old args and code them to set the new args. and have a deprecation cycle. Definitely doable.

I agree that this change is welcome, and the proposed API here is indeed much better than the previous one. However, I think this can be done better by having, as you mentioned, some deprecation warnings while still maintaining backwards compatibility for some time.

patrickvonplaten commented:
@LysandreJik - I'm fine with keeping the old arguments with deprecation warnings - will add this to the PR.

patrickvonplaten commented on src/transformers/benchmark/benchmark_args.py:
```diff
@@ -34,6 +34,34 @@
 @dataclass
 class PyTorchBenchmarkArguments(BenchmarkArguments):
+
+    deprecated_args = [
```
This keeps full backwards compatibility @LysandreJik

patrickvonplaten commented on examples/benchmarking/run_benchmark.py:
```diff
@@ -20,7 +20,25 @@
 def main():
     parser = HfArgumentParser(PyTorchBenchmarkArguments)
-    benchmark_args = parser.parse_args_into_dataclasses()[0]
+    try:
```
The `run_benchmark.py` script does not accept `--no_` ... arguments anymore. A very explicit error message is given on how the command line args have to be changed. This break backwards compatibility **only** for a command line command and only for the example files, but states explicitly how to change the input name. 
I decided to go for this option because:
a) keeping full backwards compatibility for the example file would have required to do some ugly hacks in the hfArgParser which I don't think is a good idea
b) This only concerns command line commands of example files, so I highly doubt that this change bothers anybody.

LysandreJik reviewed (approved):
Great, thanks a lot for taking the time to do this. LGTM!
